safehttp/task.go

```diff
@@ -0,0 +1,264 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package safehttp
+
+import (
+	"fmt"
+	"net/http"
+)
+
+// TODO(kele): come up with a better name than this.
+type task struct {
+	rw  http.ResponseWriter
+	req *IncomingRequest
+
+	cfg HandlerConfig
+
+	code   StatusCode
+	header Header
+
+	// TODO(kele): we need to distinguish between calling Write and actually
+	// writing to the net/http.ResponseWriter.
+	written      bool
+	writtenError bool
+}
+
+// DeprecatedNewResponseWriter creates a ResponseWriter implementation that has
+// been historically used for testing. DO NOT USE this function, it's going
+// to be removed soon.
+//
+// The main problem with this is the implementation of the returned
+// ResponseWriter is incomplete, i.e. lacks the Handler and Interceptors fields
+// of the HandlerConfig needed by the task. For the existing tests that use it,
+// it's fine, but they should be migrated.
+//
+// TODO(kele): remove this once we have a better option to provide a
+// ResponseWriter implementation for interceptor testing.
+func DeprecatedNewResponseWriter(rw http.ResponseWriter, dispatcher Dispatcher) ResponseWriter {
+	return &task{
+		cfg:    HandlerConfig{Dispatcher: dispatcher},
+		rw:     rw,
+		header: newHeader(rw.Header()),
+	}
+}
+
+// HandlerConfig is the safe HTTP handler configuration, including the
+// dispatcher and interceptors.
+type HandlerConfig struct {
+	Handler      Handler
+	Dispatcher   Dispatcher
+	Interceptors []ConfiguredInterceptor
+}
+
+func processRequest(cfg HandlerConfig, rw http.ResponseWriter, req *http.Request) {
```

Comment:
personal taste: I'd make processRequest a method of HandlerConfig.

Reply:
I did intend to keep `HandlerConfig` to be just data. Maybe I should just make it a value receiver method, though...